From 1832af6865180c636f822f1bcfcc2e55ed4b3b39 Mon Sep 17 00:00:00 2001
From: Alan Shearer <sakison@gmail.com>
Date: Fri, 25 Sep 2026 14:27:42 -0500
Subject: [PATCH] virtio_net: destroy TX/RX mutexes if interrupt setup fails

We init rxLock and txLock, then call setup_interrupt. If that
fails we jumped to err6 and never destroyed the mutexes.

err6 is also used before those mutexes exist, so I added err7
and destroy them there.

Change-Id: I166bb2f012cd4fc8c787b227a0befc77f43808ba
Reviewed-on: https://review.haiku-os.org/c/haiku/+/11858
Tested-by: Commit checker robot <no-reply+buildbot@haiku-os.org>
Reviewed-by: waddlesplash <waddlesplash@gmail.com>
Haiku-Format: Haiku-format Bot <no-reply+haikuformatbot@haiku-os.org>
---

diff --git a/src/add-ons/kernel/drivers/network/ether/virtio/virtio_net.cpp b/src/add-ons/kernel/drivers/network/ether/virtio/virtio_net.cpp
index b33bdaa..67ed27b 100644
--- a/src/add-ons/kernel/drivers/network/ether/virtio/virtio_net.cpp
+++ b/src/add-ons/kernel/drivers/network/ether/virtio/virtio_net.cpp
@@ -434,21 +434,21 @@
 	status = info->virtio->setup_interrupt(info->virtio_device, NULL, info);
 	if (status != B_OK) {
 		ERROR("interrupt setup failed (%s)\n", strerror(status));
-		goto err6;
+		goto err7;
 	}
 
 	status = info->virtio->queue_setup_interrupt(info->rxQueues[0],
 		virtio_net_rxDone, info);
 	if (status != B_OK) {
 		ERROR("queue interrupt setup failed (%s)\n", strerror(status));
-		goto err6;
+		goto err7;
 	}
 
 	status = info->virtio->queue_setup_interrupt(info->txQueues[0],
 		virtio_net_txDone, info);
 	if (status != B_OK) {
 		ERROR("queue interrupt setup failed (%s)\n", strerror(status));
-		goto err6;
+		goto err7;
 	}
 
 	if ((info->features & VIRTIO_NET_F_CTRL_VQ) != 0) {
@@ -456,13 +456,16 @@
 			NULL, info);
 		if (status != B_OK) {
 			ERROR("queue interrupt setup failed (%s)\n", strerror(status));
-			goto err6;
+			goto err7;
 		}
 	}
 
 	*_cookie = info;
 	return B_OK;
 
+err7:
+	mutex_destroy(&info->rxLock);
+	mutex_destroy(&info->txLock);
 err6:
 	for (int i = 0; i < info->txSizes[0]; i++)
 		delete info->txBufInfos[i];
